### PR TITLE
Tests for k8s.io/apimachinery/pkg/api/validate/...

### PR DESCRIPTION
**Pull request by thockin**
Thinking about how to get this upstream, I spent a bit of time on tests for k8s.io/apimachinery/pkg/api/validate/...

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/validate/net.go [hunk 1]
```diff
@@ -17,20 +17,29 @@ limitations under the License.
...
 	"k8s.io/apimachinery/pkg/api/operation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	netutils "k8s.io/utils/net"
 )
 
-// IP verifies that the specified value is a valid IP address.
-func IP(opCtx operation.Context, fldPath *field.Path, value, _ *string) field.ErrorList {
+// IPSloppy verifies that the specified value is a valid IP address, but allows
+// leading zeros on each octet value.  This should not be used for new APIs.
+func IPSloppy(opCtx operation.Context, fldPath *field.Path, value, _ *string) field.ErrorList {
```
Comment (yongruilin) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/net.go:29]: I might miss some context. May I know why this should not be used for new APIs?

File: staging/src/k8s.io/apimachinery/pkg/api/validate/enum_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		value: "a",
+		valid: sets.New("a", "b", "c"),
+		err:   false,
+	}, {
+		value: "a",
+		valid: sets.New("c", "a", "b"),
+		err:   false,
+	}, {
+		value: "a",
+		valid: sets.New("b", "c", "a"),
```
Comment (yongruilin) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/enum_test.go:42]: I don't see the point for putting different orders here for a `Set`. And how about adding a case for string alias, like `var foo string`?

**Review by yongruilin — COMMENTED** [review]
LGTM with 2 questions.

File: staging/src/k8s.io/apimachinery/pkg/api/validate/net.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/net.go:29]: Go (and therefore k8s) used to allow strings like "001.002.003.004" to parse into IP addresses.  Kubernetes generally doesn't mutate user input, so we stored that in etcd.  The problem is that tools OUTSIDE of Go handle those in different ways.  Some reject them.  Some treat them as decimal.  Some treat them as octal.

The example I gave is not a big deal, but consider "010.020.030.040" -- is that "10.20.30.40" or is it "8.16.24.32" ?  Different tools disagree.

So Go broke compat and disallowed those values.  But we have them saved on disk!  If we just use Go's validation, we we fail to update objects which we previously allowed.  That's pretty high up on the list of things we don't want to do.

So we forked Go's old parser and dubbed it "sloppy".  We allow sloppy input in existing APIs but for new APis, by definition we do not have saved data, so we should not be sloppy.

File: staging/src/k8s.io/apimachinery/pkg/api/validate/enum_test.go (hunk 2 above)
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/enum_test.go:42]: You are right.  Initially I had the input as lists, then changed it to set.  Fixed in next push.

File: staging/src/k8s.io/apimachinery/pkg/api/validate/net.go (hunk 1 above)
Comment (yongruilin) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/net.go:29]: Good to know, thanks!

**Pull request merged**